### PR TITLE
Fix analyzer issues

### PR DESCRIPTION
**Pull request by pshao25**
Fix https://github.com/Azure/autorest.csharp/issues/3710
Root cause: https://github.com/Azure/autorest.csharp/issues/3710#issuecomment-1707840791
Quote here:
> The get sub client method returns the sub client called `Operation`, which is a criterion of a service method in analyzer
> ```C#
> public virtual Operation GetOperationClient(string apiVersion = "1.0.0")
> {
>     Argument.AssertNotNull(apiVersion, nameof(apiVersion));
> 
>     return new Operation(ClientDiagnostics, _pipeline, _endpoint, apiVersion);
> }
> ```
Solution: When checking whether a method is a service method, not just checking the type name, but also checking the namespace name.

Fix https://github.com/Azure/autorest.csharp/issues/3708
Root cause: https://github.com/Azure/autorest.csharp/issues/3708#issuecomment-1788716281
Quote here: 

> We generate static properties in clients, so there will be a static constructor without any parameters implicitly added, which should not be in the searching scope of a constructor equivalent.

Solution: exclude static constructor when looking for a equivalent constructor method, because we only care about the constructors explicitly written by generator or customization.

Preview in azure-sdk-for-net: https://github.com/Azure/azure-sdk-for-net/pull/39672

### CODE REVIEW

**Review by jsquire — CHANGES_REQUESTED** [review]
Can we please add details to the description.   What are the intended changes to the current ruleset?  The code itself has few comments as to the intent of the rule and what pattern its intending to validate.

**Comment by pshao25** [comment]
> Can we please add details to the description. What are the intended changes to the current ruleset? The code itself has few comments as to the intent of the rule and what pattern its intending to validate.

Updated.

**Comment by pshao25** [comment]
@m-nash could you help review? This blocks we change constructors behaviors as we discussed in https://github.com/Azure/autorest.csharp/issues/3706#issuecomment-1752374728

File: src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0002Tests.cs [hunk 1]
```diff
@@ -294,6 +294,25 @@ public virtual Response Get(string s, CancellationToken cancellationToken)
...
+}";
+            await Verifier.CreateAnalyzer(code)
+                .RunAsync();
+        }
+
+        [Fact]
+        public async Task AZC0002NotProducedForGetSubClientMethods()
+        {
+            const string code = @"
+namespace RandomNamespace
```
Comment (AlexanderSher) [inline src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers.Tests/AZC0002Tests.cs:306]: Can we add cases with `RandomNamespace.Azure` and `Azure.RandomNamespace`?

File: src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientMethodsAnalyzer.cs [hunk 2]
```diff
@@ -156,7 +158,7 @@ bool IsValidPageable(ITypeSymbol typeSymbol)
                 }
 
                 var pageableReturn = pageableTypeSymbol.TypeArguments.Single();
-                if (!IsOrImplements(pageableReturn, BinaryDataTypeName))
+                if (!IsOrImplements(pageableReturn, BinaryDataTypeName, SystemNamespace))
```
[the remark this answers is not in the document]
Comment (pshao25) [reply src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientMethodsAnalyzer.cs:161]: If `My.System.BinaryData` extends `System.BinaryData` then it's still `true` else `false`.

[the remark this answers is not in the document]
Comment (AlexanderSher) [reply src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientMethodsAnalyzer.cs:161]: Can you please add a test for it?

**Review by AlexanderSher — APPROVED** [review]

**Pull request merged**